### PR TITLE
V4.0 ext3 supp 2

### PR DESCRIPTION
**Pull request by aviviadi**
http://issues.hibernatingrhinos.com/issue/RavenDB-5522

### CODE REVIEW

File: src/Voron/Platform/Posix/AllowBuildOnWindows.cs [hunk 1]
```diff
@@ -114,6 +114,36 @@ public static long sysconf(SysconfName name)
...
+            var orig = lseek(fd, 0, WhenceFlags.SEEK_CUR);
+            var offset = lseek(fd, size - 1, WhenceFlags.SEEK_SET);
+            if (offset == -1)
+                return offset;
+
+            int zero = 0;
+
+            int rc = (int) write(fd, &zero, 1UL);
+            if (rc == -1)
+                return rc;
```
Comment (ayende) [inline src/Voron/Platform/Posix/AllowBuildOnWindows.cs:136]: You have to return to the right location first.

File: src/Voron/StorageEnvironment.cs [hunk 2]
```diff
@@ -114,6 +128,60 @@ public StorageEnvironment(StorageEnvironmentOptions options)
...
+                }
+
+                var result = Syscall.posix_fallocate(fd, IntPtr.Zero, (UIntPtr)(64L * 1024));
+                if (result == (int)Errno.EINVAL)
+                {
+                    if (_log.IsInfoEnabled)
+                        _log.Info(
+                        "Cannot fallocate (rc = EINVAL) to a file '" + filename + "' opened using O_DIRECT. Assuming O_DIRECT is not supported by this file system");
+
+                    return false;
```
Comment (ayende) [inline src/Voron/StorageEnvironment.cs:156]: You leaked the file handle here.

File: src/Voron/StorageEnvironment.cs [hunk 3]
```diff
@@ -114,6 +128,60 @@ public StorageEnvironment(StorageEnvironmentOptions options)
...
+                {
+                    if (_log.IsInfoEnabled)
+                        _log.Info("Failed to close test file at '" + filename + "'. (rc = " + result + ").");
+                }
+            }
+            finally 
+            {
+                try
+                {
+                    File.Delete(filename);
```
Comment (ayende) [inline src/Voron/StorageEnvironment.cs]: Why not use unlink here?

File: src/Voron/StorageEnvironmentOptions.cs [hunk 4]
```diff
@@ -22,6 +22,7 @@ public abstract class StorageEnvironmentOptions : IDisposable
         public string TempPath { get; }
 
         public event EventHandler<RecoveryErrorEventArgs> OnRecoveryError;
+        public event EventHandler<NonDurabalitySupportEventArgs> OnNonDurabalitySupportError = (sender, args) => { };
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs]: This shouldn't have  default handler

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/Platform/Posix/AllowBuildOnWindows.cs (hunk 1 above)
Comment (aviviadi) [reply src/Voron/Platform/Posix/AllowBuildOnWindows.cs:136]: it is fatal error any how..

File: src/Voron/StorageEnvironment.cs (hunk 2 above)
Comment (aviviadi) [reply src/Voron/StorageEnvironment.cs:156]: will be fixed

File: src/Voron/StorageEnvironment.cs (hunk 3 above)
Comment (aviviadi) [reply src/Voron/StorageEnvironment.cs]: we don't want the user to see the file at all

File: src/Voron/StorageEnvironmentOptions.cs (hunk 4 above)
Comment (aviviadi) [reply src/Voron/StorageEnvironmentOptions.cs]: Where to add an handler for this one ? (and what to do inside of it?)

File: src/Voron/StorageEnvironment.cs [hunk 5]
```diff
@@ -159,23 +161,20 @@ private bool IsStorageSupportingO_Direct(string path)
...
-                try
-                {
-                    File.Delete(filename);
-                }
-                catch(Exception ex)
+
+                if (Syscall.unlink(filename) != 0)
                 {
-                    _log.Info("Failed to delete test file at '" + filename + "'", ex);
+                    _log.Info("Failed to delete test file at '" + filename + "'", new IOException("Failed to unlink " + filename));
```
Comment (ayende) [inline src/Voron/StorageEnvironment.cs:177]: You are missing the _reason_ for the error.

File: src/Voron/StorageEnvironmentOptions.cs [hunk 6]
```diff
@@ -43,8 +44,10 @@ public void InvokeNonDurabalitySupportError(object sender, string message, Excep
             var handler = OnNonDurabalitySupportError;
             if (handler == null)
             {
-                throw new InvalidDataException(message + Environment.NewLine +
-                     "An exception has been thrown because there isn't a listener to the OnNonDurabalitySupportError event on the storage options.", e);
+                if (_log.IsInfoEnabled)
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs:47]: No, it _should_ throw if no one is actually listening. This is a problem that we will handle in `ServerStore`

**Review by ayende — CHANGES_REQUESTED** [review]

**Pull request merged**